proposals/dnd-list.md

```diff
@@ -0,0 +1,133 @@
+# Drag and Drop List
+
+## Usage
+
+### Basic usage
+
+```js
+import * as React from 'react';
+import {StatefulList, Item} from 'baseui/dnd-list';
+
+export default () =>
+  <StatefulList
+    initialState={{items: ['Item 1', 'Item 2', 'Item 3']}}
+    onChange={({newState, oldIndex, newIndex}) => {}}
+  />
+```
+
+### Advanced usage
+
+```js
+import * as React from 'react';
+import {
+  StatefulList,
+  StyledList,
+  StyledItem
+} from 'baseui/dnd-list';
+import {styled} from 'baseui';
+
+CustomList = styled(StyledList, props => ({
+  border: '1px solid green'
+}));
+
+CustomItem = styled(StyledItem, props => ({
+  borderBottom: props.$dragged ? `1px solid red` : 'none'
+}));
+
+export default () =>
+  <StatefulList
+    initialState={{ items: ['Item 1', 'Item 2', 'Item 3']}}
+    overrides={{
+      List: CustomList,
+      Item: CustomItem
+    }}
+  >
+```
+
+## Exports
+
+* `StatefulList`
+* `List`
+* `StyledList`
+* `StyledItem`
+* `StyledDragHandle`
+* `StyledCloseHandle`
+* `STATE_CHANGE_TYPE`
+
+## `List` API
+
+* `items: Array<string>` - Required.
+  List items.
+* `overrides: {Root, Item, DragHandle, CloseHandle}` - Optional.
```

Comment:
what do we assume the `CloseHandle` button does? Does there need to be some api for removing items? I think this should be not implemented initially, until there is some time for folks to use the component. Does a parent/child api better serve customization?

```jsx
<StatefulList>
  {items.map(x => <Item>{x}</Item>}
</StatefulList>

<StatefulList>
  {items.map(x => (
    <Item>
      {x}
      <Button onClick={...remove item from list}>
        remove
      </Button>
    </Item>
  }
</StatefulList>
```

Reply:
> what do we assume the CloseHandle button does? Does there need to be some api for removing items?

There is an `x` in the mocks so I assumed it means "remove". Technically, it would mean moving the item from index `i` to `-1` so there would be no need for an extra API.

> Does a parent/child api better serve customization?

Good point. I felt we could always do this later if there's a good use-case. One could be having differently styled items but not sure how realistic is that.

On the other hand, this API would introduce some footguns since `List` needs to pass some props to `Items` through `React.cloneElement`, so if `Item` is a custom component it could be easily broken if those hidden props are not correctly passed down.

Reply:
Cool I'm all for the encapsulated pattern. Though, still somewhat concerned about optimizing the api based on a possibly unique behavior rather than the default case. Will we need an opt-out flag at the top level?

Reply:
> Will we need an opt-out flag at the top level?

In respect to removing items? Or can you elaborate?

Reply:
yeah i'm thinking about the case where items should not be removed. will we need a `removable={false}` prop? Is there a way we can empower users to implement this use-case themselves and not complicate the base component?

Reply:
How about this? I can imagine we would add `disabled` as well at some point.

```jsx
// default removable = false
const items = [
  {label: 'Item 1', removable: false},
  {label: 'Item 2', removable: true},
  {label: 'Item 3', removable: false}
];

export default () =>
  <StatefulList
    initialState={{ items }}
    onChange={({newState, oldIndex, newIndex}) => {}}
  />
```